Library/Homebrew/formula.rb

```diff
@@ -2140,25 +2158,13 @@ def to_hash
       "keg_only"                 => keg_only?,
       "keg_only_reason"          => keg_only_reason&.to_hash,
       "options"                  => [],
-      "build_dependencies"       => dependencies.select(&:build?)
-                                                .map(&:name)
-                                                .uniq,
-      "dependencies"             => dependencies.reject(&:optional?)
-                                                .reject(&:recommended?)
-                                                .reject(&:build?)
-                                                .reject(&:test?)
-                                                .map(&:name)
-                                                .uniq,
-      "test_dependencies"        => dependencies.select(&:test?)
-                                                .map(&:name)
-                                                .uniq,
-      "recommended_dependencies" => dependencies.select(&:recommended?)
-                                                .map(&:name)
-                                                .uniq,
-      "optional_dependencies"    => dependencies.select(&:optional?)
-                                                .map(&:name)
-                                                .uniq,
-      "uses_from_macos"          => uses_from_macos_elements.uniq,
+      "deps"                     => [],
+      "build_dependencies"       => [], # deprecated
+      "dependencies"             => [], # deprecated
+      "test_dependencies"        => [], # deprecated
+      "recommended_dependencies" => [], # deprecated
+      "optional_dependencies"    => [], # deprecated
+      "uses_from_macos"          => [], # deprecated
```

Comment:
> * We don't supply `uses_from_macos` bounds information.

I think deprecating `uses_from_macos` would be reasonable, I'd be surprised if it is widely consumed as-is given the constraints you've mentioned 👍🏻 

> * An alternative solution could be to be to redefine these as "stable-only" dependencies

Yeh, this would make sense to me 👍🏻 

> * to provide an alternative `head_dependencies` that declares the set of dependencies again for the head spec

Similarly, 👍🏻 if we decide it's needed.

> There is the interesting question though here regardless of compatibility. We already have one field marked for deprecation: `oldname` (this one was definitely unavoidable). How do we handle this?

Good question! We can't really warn on usage. Perhaps we need to require a `brew --info --json=v3` or something? Alternatively, we just document it's going away, remove the data in the next release, remove the field in the release after?

> Beyond that, it's not clear. We have `brew info --json=v1` and `v2`. The differences are on a higher level rather than `Formula#to_hash`, but we never have deprecated v1 there.

Indeed. Deprecating v1 would make sense definitely.

> I agree that JWS isn't yet documented on formulae.brew.sh so is considered private API for now and we can change it whenever we want - though older Homebrew versions may be a concern.

I think given the way Homebrew will update itself and we encourage people to run `brew update` when things break: we don't need to care about old versions in this way.

I'd also love it if we kept this undocumented forever to provide a firewall between "public API" and "private API we consume" so the latter can be more minimal and changeable.

Reply:
> I think deprecating `uses_from_macos` would be reasonable, I'd be surprised if it is widely consumed as-is given the constraints you've mentioned 👍🏻

The question then what would be its replacement. The replacement was a combined `deps` for everything, but if we don't do that and go the `head_dependencies` route then we'll still need something for `uses_from_macos` dependencies.

Reply:
`uses_from_macos_dependencies`, I guess?

To be clear, I'm still not opposed to a decent refactor of the output and to have a `deps` field. I think at that point it may be worth doing a full, ground-up rebuild of our JSON to improve the naming/nesting and also potentially have a "slim" version of only what we use/need for the API installation and "full" version with all the data we need.

Reply:
Ok what I'm thinking (assuming it works) is to add `head_dependencies` and to keep `uses_from_macos` but add an additional `uses_from_macos_bounds` for now.

I will then create (but not push to this PR) a commit that refactors all that into a combined `deps` field, that I'll keep around in a fork branch and we can cherry-pick should in the future go ahead with a bigger rebuild of the JSON (v3 or whatever we call it), potentially also aligning any differences we can between formulae and casks. I will link that commit somewhere here probably.

Reply:
Cool, thanks, that works great 👍🏻